File: ui/src/lib/types/climb.ts
import z from "zod";
import type { API } from "./api";
import { JSONBody } from "./general";
import type { Gym, Source } from "./gym";
import { convertGym } from "./gym";

export enum ClimbType {
  Boulder = "boulder",
  Lead = "lead",
}

export enum ClimbFinish {
  Flash = "flash",
  Top = "top",
  Repeat = "repeat",
}

export interface Climb {
  id: number;
  grade: number;
  holdColor: string;
  climbType: ClimbType;
  finishType: ClimbFinish;
  source: Source;
}

export interface ClimbDay {
  id: number;
  date: Date;
  gym: Gym;
  climbs: Climb[];
  source: Source;
}

export interface ClimbStatsProgress {
  date: string;
  grade: number;
  volume: number;
}

export interface ClimbStatsGrade {
  grade: number;
  flash: number;
  top: number;
  repeat: number;
}

export interface ClimbStats {
  total: number;
  totalUnique: number;
  flash: number;
  top: number;
  repeat: number;
  best: number;
  bestAmount: number;
  bestFlash: number;
  bestFlashAmount: number;
  sessions: number;
  boulder: number;
  lead: number;
  medianClimbsPerSession: number;
  graphProgress: ClimbStatsProgress[];
  graphPerGrade: ClimbStatsGrade[];
}

// Converts

export const convertClimb = (c: API.Climb): Climb => ({
  id: c.id,
  grade: c.grade,
  holdColor: c.hold_color,
  climbType: c.climb_type as ClimbType,
  finishType: c.finish_type as ClimbFinish,
  source: c.source as Source,
});

export const convertClimbs = (c: API.Climb[]): Climb[] => c.map(convertClimb);

export const convertClimbDay = (d: API.ClimbDay): ClimbDay => ({
  id: d.id,
  date: new Date(d.date),
  gym: convertGym(d.gym),
  climbs: convertClimbs(d.climbs),
  source: d.source as Source,
});

export const convertClimbDays = (d: API.ClimbDay[]): ClimbDay[] => d.map(convertClimbDay);

export const convertClimbStats = (s: API.ClimbStats): ClimbStats => ({
  total: s.total,
  totalUnique: s.total_unique,
  flash: s.flash,
  top: s.top,
  repeat: s.repeat,
  best: s.best,
  bestAmount: s.best_amount,
  bestFlash: s.best_flash,
  bestFlashAmount: s.best_flash_amount,
  sessions: s.sessions,
  boulder: s.boulder,
  lead: s.lead,
  medianClimbsPerSession: s.median_climbs_per_session,
  graphProgress: s.graph_progress,
  graphPerGrade: s.graph_per_grade,
});

export const convertClimbCreateSchema = (c: Climb): ClimbCreate => ({
  _clientId: crypto.randomUUID(),
  grade: c.grade,
  holdColor: c.holdColor,
  climbType: c.climbType,
  finishType: c.finishType,
})

export const convertClimbDayUpdateSchema = (c: ClimbDay): ClimbDayUpdate => ({
  id: c.id,
  date: c.date,
  gymId: c.gym.id,
  climbs: c.climbs.map(convertClimbCreateSchema),
})

// Schemas

export const climbCreateSchema = z.object({
  _clientId: z.string(),
  grade: z.number().positive(),
  holdColor: z.string(),
  climbType: z.enum(ClimbType),
  finishType: z.enum(ClimbFinish),
})
export type ClimbCreate = z.infer<typeof climbCreateSchema> & JSONBody

export const climbDayCreateSchema = z.object({
  date: z.date(),
  gymId: z.number().positive(),
  climbs: z.array(climbCreateSchema).min(1),
})
export type ClimbDayCreate = z.infer<typeof climbDayCreateSchema> & JSONBody

export const climbDayUpdateSchema = z.object({
  id: z.number().positive(),
  date: z.date(),
  gymId: z.number().positive(),
  climbs: z.array(climbCreateSchema).min(1),
})
export type ClimbDayUpdate = z.infer<typeof climbDayUpdateSchema> & JSONBody

File: internal/server/dto/climb.go
package dto

import (
	"time"

	"github.com/Topvennie/beta-log/internal/database/model"
	"github.com/Topvennie/beta-log/internal/grade"
	"github.com/Topvennie/beta-log/pkg/utils"
)

type climb struct {
	ID         int               `json:"id"`
	Grade      string            `json:"grade"`
	HoldColor  string            `json:"hold_color"`
	ClimbType  model.ClimbType   `json:"climb_type"`
	FinishType model.ClimbFinish `json:"finish_type"`
	Source     model.Source      `json:"source"`
}

func climbDTO(c *model.Climb, setting model.Setting) climb {
	return climb{
		ID:         c.ID,
		Grade:      grade.Grade(c.Grade).Format(setting.GradeSystem),
		HoldColor:  c.HoldColor,
		ClimbType:  c.ClimbType,
		FinishType: c.FinishType,
		Source:     c.Source,
	}
}

type ClimbDay struct {
	ID     int          `json:"id"`
	Date   time.Time    `json:"date"`
	Gym    Gym          `json:"gym"`
	Climbs []climb      `json:"climbs"`
	Source model.Source `json:"source"`
}

func ClimbDayDTO(d *model.ClimbDay, setting model.Setting) ClimbDay {
	return ClimbDay{
		ID:     d.ID,
		Date:   d.Date,
		Gym:    GymDTO(&d.Gym),
		Climbs: utils.SliceMap(d.Climbs, func(c model.Climb) climb { return climbDTO(&c, setting) }),
		Source: d.Source,
	}
}

type ClimbDayFilter struct {
	UserID int
	Limit  int
	Offset int
}

func (c ClimbDayFilter) ToModel() model.ClimbDayFilter {
	return model.ClimbDayFilter(c)
}

type ClimbCreate struct {
	Grade      string            `json:"grade"`
	HoldColor  string            `json:"hold_color" validate:"hexcolor"`
	ClimbType  model.ClimbType   `json:"climb_type" validate:"required"`
	FinishType model.ClimbFinish `json:"finish_type" validate:"required"`
}

func (c ClimbCreate) ToModel(setting model.Setting) model.Climb {
	return model.Climb{
		Grade:      int(grade.FromString(c.Grade, setting.GradeSystem)),
		HoldColor:  c.HoldColor,
		ClimbType:  c.ClimbType,
		FinishType: c.FinishType,
	}
}

type ClimbDayCreate struct {
	Date   time.Time     `json:"date" validate:"required"`
	GymID  int           `json:"gym_id" validate:"required,min=1"`
	Climbs []ClimbCreate `json:"climbs" validate:"required,min=1,dive"`
}

func (c ClimbDayCreate) ToModel(setting model.Setting) model.ClimbDay {
	return model.ClimbDay{
		Date:   c.Date,
		GymID:  c.GymID,
		Climbs: utils.SliceMap(c.Climbs, func(c ClimbCreate) model.Climb { return c.ToModel(setting) }),
	}
}

type ClimbDayUpdate struct {
	ID     int           `json:"id" validate:"required"`
	Date   time.Time     `json:"date" validate:"required"`
	GymID  int           `json:"gym_id" validate:"required,min=1"`
	Climbs []ClimbCreate `json:"climbs" validate:"required,min=1,dive"`
}

func (c ClimbDayUpdate) ToModel(setting model.Setting) model.ClimbDay {
	return model.ClimbDay{
		ID:     c.ID,
		Date:   c.Date,
		GymID:  c.GymID,
		Climbs: utils.SliceMap(c.Climbs, func(c ClimbCreate) model.Climb { return c.ToModel(setting) }),
	}
}

type ClimbStatsProgress struct {
	Date   string `json:"date"`
	Grade  string `json:"grade"`
	Volume int    `json:"volume"`
}

type ClimbStatsGrade struct {
	Grade  string `json:"grade"`
	Flash  int    `json:"flash"`
	Top    int    `json:"top"`
	Repeat int    `json:"repeat"`
}

type ClimbStats struct {
	Total                  int                  `json:"total"`
	TotalUnique            int                  `json:"total_unique"`
	Flash                  int                  `json:"flash"`
	Top                    int                  `json:"top"`
	Repeat                 int                  `json:"repeat"`
	Best                   string               `json:"best"`
	BestAmount             int                  `json:"best_amount"`
	BestFlash              string               `json:"best_flash"`
	BestFlashAmount        int                  `json:"best_flash_amount"`
	Sessions               int                  `json:"sessions"`
	Boulder                int                  `json:"boulder"`
	Lead                   int                  `json:"lead"`
	MedianClimbsPerSession int                  `json:"median_climbs_per_session"`
	GraphProgress          []ClimbStatsProgress `json:"graph_progress"`
	GraphPerGrade          []ClimbStatsGrade    `json:"graph_per_grade"`
}
